### PR TITLE
OMPI/OSHMEM: add new functionality of OpenSHMEM v1.4.

### PR DESCRIPTION
**Pull request by xinzhao3**
This PR contains following new functionality of OpenSHMEM 1.4:

1. context object (RMA, AMO, fence/quiet)
2. thread support
3. shmem_calloc
4. shmem_test

bitwise atomics and shmem_sync will be added soon.

### CODE REVIEW

**Comment by jladd-mlnx** [comment]
👍 still need to modify Ikrit

**Comment by jladd-mlnx** [comment]
@hoopoepg @xinzhao3 is this ready to merge?

**Comment by hoopoepg** [comment]
@jladd-mlnx I think yes
but there are git conflicts which should be resolved

**Comment by jladd-mlnx** [comment]
@xinzhao3 please resolve the conflicts so we can merge this.

**Comment by hppritcha** [comment]
@xinzhao3 if you could get this in this week that would be great.

**Comment by hppritcha** [comment]
@xinzhao3 need to rebase this soon or 4.0 will leave the station without OpenSHMEM 1.4.
@jladd-mlnx

File: oshmem/include/shmem.h.in [hunk 1]
```diff
@@ -452,6 +758,28 @@ OSHMEM_DECLSPEC long long shmem_longlong_cswap(long long *target, long long cond
 #endif
 
 /* Atomic Fetch&Add */
+OSHMEM_DECLSPEC int shmem_ctx_int_atomic_fadd(shmem_ctx_t ctx, int *target, int value, int pe);
```
Comment (hoopoepg) [inline oshmem/include/shmem.h.in]: BTW, here should be naming ```shmem_<TYPENAME>_atomic_fetch_add```

**Comment by jladd-mlnx** [comment]
@hppritcha @gpaulsen - pending one last review from @hoopoepg this is ready to go. OMPI master will be OSHMEM v1.4 compliant upon merging. 

@hoopoepg could you please give one final review.

**Comment by xinzhao3** [comment]
need some changes on bitwise operations and context objects. Will push soon.

**Comment by gpaulsen** [comment]
@xinzhao3 @jladd-mlnx Can you please prepare this branch for merging today (if possible), and get into master if possible?  We can fix bugs on the v4.0.x branch as needed, but it sounds like most of this functionality is ready...  Since this is a large feature for v4.0.x I'd like to delay branching v4.0.x until this is merged to master.

**Comment by jladd-mlnx** [comment]
@gpaulsen We will have it merged today.

**Comment by gpaulsen** [comment]
+1

**Pull request merged**